add an API to retrieve an enum's `name` property

leiflm opened the pull request:
Enums may have a `name` that is typically displayed to the user. This adds an API to allow access to that.

leiflm committed:
add an API to retrieve an enum's `name` property

Enums may have a `name` that is typically displayed to the user.
This adds an API to allow access to that.

leelavg commented on src/pylero/base_polarion.py:
```diff
@@ -1061,6 +1061,36 @@ def get_valid_field_values(self, enum_id, control=None):
+            self._cache["enums"][enum_id] = {}
+            self._cache["enums"][enum_id][control] = enums
+        # the _cache contains _suds_object, so the id attribute is used.
+        return next([enum.name for enum in enums if enum.id == value])
```
the code is similar to `get_valid_field_values` which looks fine but I don't understand the intention behind `return next(list)`.
- if you are looking for a generator then it should be `n = iter(list); yield next(n)`, however I don't get why you need a generator expression though 🤔 

nevertheless, my python skills are becoming rusty, correct me if you are trying something else.

thanks!

leiflm committed:
fixup: to not use `next`

leelavg commented on src/pylero/base_polarion.py:
```diff
@@ -1061,6 +1061,42 @@ def get_valid_field_values(self, enum_id, control=None):
+        # the _cache contains _suds_object, so the id attribute is used.
+        if items := [enum.name for enum in enums if enum.id == value]:
+            if len(items) > 0:
+                return items[0]
```
if only interested in first item there is no need for list comprehension, you could do
```suggestion
        for enum in enums:
          if enum.id == value:
                return enum.name
```

leiflm committed:
Update src/pylero/base_polarion.py

adapt suggested, more untuitive solution instead of unnecessary list comprehension.

Co-authored-by: Leela Gangavarapu <leelavg@thoughtexpo.com>

waynesun09 reviewed (changes requested):
## Review Summary

Thanks for the contribution — the intent is solid. Having an API to resolve an enum value to its display name is a natural complement to `get_valid_field_values`. However, there are a couple of bugs that need fixing before merge, plus some improvements worth considering.

---

### Bugs (must fix)

**1. Broken f-string interpolation in error message**

The second string literal is missing the `f` prefix, so `{value}` renders literally instead of being interpolated. There's also a missing space before "or":

```python
# Current (broken):
raise PyleroLibException(
        f'The enum element "{enum_id}"'
        'or its value "{value}" do not exist.'
    )

# Fixed:
raise PyleroLibException(
        f'The enum element "{enum_id}" '
        f'or its value "{value}" do not exist.'
    )
```

**2. Inconsistent indentation in the `for` loop body**

The `if` statement uses 2-space indentation (10 spaces total) instead of the project-standard 4-space increment (12 spaces total). This will likely fail the pre-commit linter checks:

```python
# Current (wrong):
        for enum in enums:
          if enum.id == value:   # 2-space indent
                return enum.name

# Fixed:
        for enum in enums:
            if enum.id == value:   # 4-space indent
                return enum.name
```

*(This is likely why the CI "Code Quality (3.11)" job failed on the original run.)*

---

### Warnings (should fix)

**3. Docstring is inaccurate** — The description "Gets the available enumeration options" is copied from `get_valid_field_values`. This method resolves a single value to its display name. Suggested:

```python
"""Resolves an enumeration value to its display name.
...
Returns:
    The display name (enum.name) for the given value

Raises:
    PyleroLibException: If no option with the given value exists
"""
```

**4. Code duplication** — The entire cache-lookup block is copy-pasted from `get_valid_field_values`. Consider extracting a shared helper:

```python
def _get_enum_options(self, enum_id, control=None):
    """Fetch enum options for the given id, using cache."""
    project_id = getattr(self, "project_id", None) or self.default_project
    enum_base = self._cache["enums"].get(enum_id)
    enums = None
    if enum_base:
        enums = enum_base.get(control)
    if not enums:
        enums = self.session.tracker_client.service.getEnumOptionsForIdWithControl(
            project_id, enum_id, control
        )
        if enum_id not in self._cache["enums"]:
            self._cache["enums"][enum_id] = {}
        self._cache["enums"][enum_id][control] = enums
    return enums
```

Then both methods become thin wrappers — easier to maintain and test.

**5. Cache overwrite bug (pre-existing)** — `self._cache["enums"][enum_id] = {}` wipes previously cached control keys for that `enum_id`. Should be:

```python
if enum_id not in self._cache["enums"]:
    self._cache["enums"][enum_id] = {}
self._cache["enums"][enum_id][control] = enums
```

This is inherited from `get_valid_field_values`, but since you're touching this code it's a good opportunity to fix.

**6. No tests** — Consider adding tests for: happy path (value exists → name returned), not-found (raises `PyleroLibException`), and cache behavior.

---

### Nits

**7. Error message could be clearer** — By the time the exception is raised, we know the enum resolved (otherwise the WSDL call would have failed). Only the `value` was not found. A clearer message:

```python
f'Value "{value}" not found in enum "{enum_id}" (control={control!r})'
```

waynesun09 reviewed (commented):
Inline comments for findings 1-4. See the earlier review for the full summary.

leiflm committed:
Update src/pylero/base_polarion.py

The second string literal is missing the f prefix, so {value} will render literally as {value} instead of being interpolated. There's also a missing space before "or".

Co-authored-by: Wayne Sun <ericbreeze@gmail.com>

leiflm committed:
Update src/pylero/base_polarion.py

The `if` uses 2-space indent (10 spaces total) instead of the project-standard 4-space increment (12 spaces total).

Co-authored-by: Wayne Sun <ericbreeze@gmail.com>

leiflm committed:
Update src/pylero/base_polarion.py

describe what this method actually does

Co-authored-by: Wayne Sun <ericbreeze@gmail.com>

leiflm committed:
src/pylero/base_polarion.py: deduplicate code

Deduplicate code by factorizing implementations.

Co-authored-by: Wayne Sun <ericbreeze@gmail.com>

leiflm committed:
Merge branch 'main' into feature/enum_display_name_api

waynesun09 reviewed (commented):
## Re-review of updated changes

Thanks for addressing the feedback — the refactoring looks great. All critical issues from the first review are resolved:

| Previous Feedback | Status |
|---|---|
| Broken f-string (missing `f` prefix) | Fixed |
| Inconsistent 2-space indentation | Fixed |
| Inaccurate docstring | Fixed |
| Extract shared `_get_enum_options` helper | Fixed — clean extraction |
| Cache overwrite bug | Fixed (`if enum_id not in`) |

The refactoring is correct — `get_valid_field_values` preserves identical behavior, and the `_get_enum_options` helper is a clean abstraction.

---

### CI is failing — `ruff-format`

The pre-commit `ruff-format` hook is reformatting two spots. Please run `pre-commit run -a` locally and push the fix:

1. The `getEnumOptionsForIdWithControl(...)` call — ruff wants it on one line (fits within 120 chars)
2. The `raise PyleroLibException(...)` — ruff wants it on one line

---

### Remaining low-severity items (non-blocking)

**Error message could be clearer** — By the time the exception is raised, the enum resolved successfully; only the `value` wasn't found. Current message implies both might be invalid. Consider:
```python
f'Value "{value}" not found in enum "{enum_id}"'
```

**`get_valid_field_values` docstring** — It says `Returns: Array of EnumOptions` but actually returns a list of string IDs. Natural time to fix since docstrings were already touched.

**No tests** — Not a merge blocker, but a unit test for `get_name_for_field_value` (happy path + not-found) with a mocked `_get_enum_options` would be easy to add.

leiflm committed:
comply with ruff-format

Accept rather long (≤ 120 chars) lines.

waynesun09 reviewed (approved):
LGTM. The refactoring is clean, `get_valid_field_values` preserves identical behavior, the cache overwrite bug fix is a nice improvement, and CI passes.

Minor non-blocking suggestions for a follow-up:
- Error message could be more precise (the enum resolved; only the value wasn't found)
- Cache docstring is now duplicated across all three methods — could be trimmed from the two public ones
- Unit tests for `get_name_for_field_value` would be a good addition
